File: llgo/test/execution/types/recursive.go
// RUN: llgo -o %t %s
// RUN: %t 2>&1 | FileCheck %s

// CHECK: 3
// CHECK-NEXT: 4

package main

type T1 *T1

func count(t T1) int {
	if t == nil {
		return 1
	}
	return 1 + count(*t)
}

func testSelfPointer() {
	var a T1
	var b T1
	var c T1 = &b
	*c = &a
	println(count(c))
	println(count(&c))
}

func main() {
	testSelfPointer()
}

File: debuginfo-tests/dexter/feature_tests/commands/perfect/expect_step_kind/recursive.cpp
// Purpose:
//      Check that \DexExpectStepKind correctly handles recursive calls.
//      Specifically, ensure recursive calls count towards 'FUNC' and not
//      'VERTICAL_BACKWARD'.
//
// REQUIRES: system-linux, lldb
//
// RUN: %dexter_base test --fail-lt 1.0 -w  \
// RUN:     --builder 'clang' --debugger 'lldb' --cflags "-O0 -g" -- %s \
// RUN:     | FileCheck %s
// CHECK: recursive.cpp:

int func(int i) {
    if (i > 1)
        return i + func(i - 1);
    return i;
}

int main()
{
    return func(3);
}

// main, func, func, func
// DexExpectStepKind('FUNC', 4)
// DexExpectStepKind('VERTICAL_BACKWARD', 0)
